Library/src/actuators/Rudder.cpp

```diff
@@ -0,0 +1,182 @@
+/*    
+    This file is a part of Stonefish.
+
+    Stonefish is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Stonefish is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+//
+//  Rudder.cpp
+//  Stonefish
+//
+//  Created by Nils Bore on 29/01/2021.
+//  Copyright (c) 2021 Nils Bore, Patryk Cieslak. All rights reserved.
+//
+
+#include "actuators/Rudder.h"
+
+#include "core/SimulationApp.h"
+#include "core/SimulationManager.h"
+#include "graphics/GLSLShader.h"
+#include "graphics/OpenGLContent.h"
+#include "entities/SolidEntity.h"
+
+namespace sf
+{
+
+Rudder::Rudder(std::string uniqueName, SolidEntity* rudder, Scalar area, Scalar liftCoeff, Scalar dragCoeff, Scalar maxAngle, bool inverted) : LinkActuator(uniqueName)
+{
+    this->dragCoeff = dragCoeff;
+    this->liftCoeff = liftCoeff;
+    this->area = area;
+    this->maxAngle = maxAngle;
+    inv = inverted;
+
+    setpoint = Scalar(0);
+    theta = Scalar(0);
+    this->rudder = rudder;
+    this->rudder->BuildGraphicalObject();
+}
+
+Rudder::~Rudder()
+{
+    if(rudder != NULL)
+        delete rudder;
+}
+
+ActuatorType Rudder::getType()
+{
+    return ActuatorType::RUDDER;
+}
+
+void Rudder::setSetpoint(Scalar s)
+{
+    if(inv) s *= Scalar(-1);
+    setpoint = std::max(std::min(s, maxAngle), -maxAngle);
+}
+
+Scalar Rudder::getSetpoint()
+{
+    return inv ? -setpoint : setpoint;
+}
+
+Scalar Rudder::getAngle()
+{
+    return theta;
+}
+
+void Rudder::Update(Scalar dt)
+{
+    //Update rudder angle
+    theta = setpoint;
+
+    if(attach != NULL)
+    {
+        Quaternion rudderRot(theta, 0, 0);
+
+        //Get transforms
+        Transform solidTrans = attach->getCGTransform();
+        // o2a is the transform of the actuator
+        Transform rudderTrans = attach->getOTransform() * o2a * Transform(rudderRot);
+        Vector3 relPos = rudderTrans.getOrigin() - solidTrans.getOrigin();
+
+        Vector3 velocity = rudderTrans.getBasis().transpose()*attach->getLinearVelocityInLocalPoint(relPos);
```

Comment:
Is it correct that this returns the absolute velocity rather than the velocity relative to the water? That would explain the behavior I'm seeing at the moment. If this is the case, how would I get the relative velocity?

Reply:
Should I combine it with this one: https://github.com/patrykcieslak/stonefish/blob/master/Library/include/entities/forcefields/Ocean.h#L86 ?

Reply:
That's what I did now (see below).